### PR TITLE
update to new compilers /py37

### PR DESCRIPTION
**Pull request by looooo**
Checklist
* [ ] Used a fork of the feedstock to propose changes
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by looooo** [comment]
@mingwandroid @isuruf @jakirkham any ideas how how to solve OSX:

```
  No known features for CXX compiler

  "Clang"

  version 4.0.1
```

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,23 +1,26 @@
...
     sha256: {{ sha256 }}
 
 build:
-    number: 1
+    number: 2
     skip: True  # [py < 36]
 
 requirements:
     build:
         - {{ compiler('cxx') }}
```
Comment (jakirkham) [inline recipe/meta.yaml]: Maybe add `- {{ compiler('c') }}` too?

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
@jakirkham thanks for the hint, but still the same problem.

```
CMake Error in CMakeLists.txt:
  No known features for CXX compiler

  "Clang"

  version 4.0.1.
```

**Comment by jakirkham** [comment]
It seems another `CMakeLists.txt` may be generating the error. Netgen perhaps?

**Comment by looooo** [comment]
hmm, there is no CMakeLists.txt in the netgen osx-package. Maybe another CMakeLists.txt in a subdirectory?

File: recipe/build.sh [hunk 2]
```diff
@@ -2,19 +2,19 @@ mkdir -p build
...
 
 cmake -G "Ninja" \
       -D CMAKE_INSTALL_PREFIX:FILEPATH=$PREFIX \
       -D CMAKE_BUILD_TYPE=Release \
       -D CMAKE_PREFIX_PATH:FILEPATH=$PREFIX \
       -D CMAKE_SYSTEM_PREFIX_PATH:FILEPATH=$PREFIX \
-      -D CMAKE_CXX_FLAGS=$FLAGS \
+      -D CMAKE_CXX_STANDARD=${cpp_std} \
       -D ENABLE_MED:BOOL=OFF \
       -D ENABLE_NETGEN:BOOL=ON \
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply recipe/build.sh]: The alternative would be to patch in the following in the CMake files.

```cmake
cmake_policy(SET CMP0025 NEW)
```

**Comment by jakirkham** [comment]
Right, that's what I was meaning.

**Comment by looooo** [comment]
After some local iterations it seems like the mentioned problem is correlated to linking to vtk:

`target_link_libraries(SMESHControls ${VTK_LIBRARIES})`

Where ${VTK_LIBRARIES} looks like this:

```
vtksysvtkCommonCorevtkCommonMathvtkCommonMiscvtkCommonSystemvtkCommonTransformsvtkCommonDataModelvtkCommonColorvtkCommonExecutionModelvtkCommonComputationalGeometryvtkFiltersCorevtkFiltersGeneralvtkImagingCorevtkImagingFouriervtkalglibvtkFiltersStatisticsvtkFiltersExtractionvtkInfovisCorevtkFiltersGeometryvtkFiltersSourcesvtkRenderingCore$PREFIX/lib/libz.dylib$PREFIX/lib/libfreetype.dylibvtkRenderingFreeTypevtkRenderingContext2DvtkChartsCorevtkDICOMParser$PREFIX/lib/liblz4.dylibvtkIOCorevtkIOLegacy$PREFIX/lib/libexpat.dylibvtkIOXMLParservtkDomainsChemistryvtkglewvtkRenderingOpenGL2vtkDomainsChemistryOpenGL2vtkIOXMLvtkParallelCorevtkFiltersAMRvtkFiltersFlowPathsvtkFiltersGenericvtkImagingSourcesvtkFiltersHybridvtkFiltersHyperTreevtkImagingGeneralvtkFiltersImagingvtkFiltersModelingvtkFiltersParallelvtkFiltersParallelImagingvtkFiltersPointsvtkFiltersProgrammable$PREFIX/lib/libpython3.7m.dylibvtkWrappingToolsvtkWrappingPythonCorevtkFiltersPythonvtkFiltersSMPvtkFiltersSelectionvtkFiltersTexturevtkFiltersTopologyverdictvtkFiltersVerdictvtkmetaio$PREFIX/lib/libjpeg.dylib$PREFIX/lib/libpng.dylib$PREFIX/lib/libtiff.dylibvtkIOImagevtkImagingHybridvtkInfovisLayoutvtkInteractionStylevtkImagingColorvtkRenderingAnnotationvtkRenderingVolumevtkInteractionWidgetsvtkViewsCorevtkproj4vtkGeovisCore$PREFIX/lib/libhdf5.dylib/usr/lib/libpthread.dylib/usr/lib/libdl.dylib/usr/lib/libm.dylib$PREFIX/lib/libhdf5_hl.dylibvtkIOAMRvtkIOEnSight$PREFIX/lib/libnetcdf.dylibvtkexoIIcvtkIOExodusvtkgl2psvtkRenderingGL2PSOpenGL2vtklibharuvtkIOExportvtkIOExportOpenGL2vtkIOGeometryvtkIOImport$PREFIX/lib/libxml2.dylibvtkIOInfovisvtkIOLSDynavtkIOMINC$PREFIX/lib/libtheoraenc.dylib$PREFIX/lib/libtheoradec.dylib$PREFIX/lib/libogg.dylibvtkIOMovievtknetcdfcppvtkIONetCDFvtkIOPLY$PREFIX/lib/libjsoncpp.dylibvtkIOParallelvtkIOParallelXMLvtksqlitevtkIOSQLvtkIOTecplotTablevtkIOVideovtkImagingMathvtkImagingMorphologicalvtkImagingStatisticsvtkImagingStencilvtkInteractionImagevtkPythonInterpretervtkRenderingContextOpenGL2vtkRenderingImagevtkRenderingLODvtkRenderingLabelvtkRenderingMatplotlibvtkRenderingVolumeOpenGL2vtkViewsContext2DvtkViewsInfovis
```

But I still have no idea what this is about.

**Comment by looooo** [comment]
To me it seems there are two bugs involved:

1. cmake is reporting nonsense instead of library <name> can not be found
This can be reproduced by adding this command:
`target_link_library(<lib> libReallyNonExisting)`

2. vtkLibraries is missing the version suffix:
`target_link_library(<lib> vtkCommonCore)` # not working
`target_link_library(<lib> vtkCommonCore-8.1.2)` # works

**Comment by looooo** [comment]
@mingwandroid is it possible this is related to the new-compilers? I have the impression that it is not possible to link to symlinks:
`target_link_library(<lib> vtkCommonCore)` doesn't work
`target_link_library(<lib> vtkCommonCore.1)` works

with `libvtkCmmonCore.dylib` beeing a link to `libvtkCommonCore.1.dylib`

maybe you have an idea what is going on here?

**Comment by mingwandroid** [comment]
> is it possible this is related to the new-compilers

I do not have time to look into that, anything is possible. You need to figure out a way to determine this for yourself though.

**Comment by looooo** [comment]
> I do not have time to look into that, anything is possible. You need to figure out a way to determine this for yourself though.

Yes it's a problem with cmake not detecting the compiler-features. But finding the root of the problem is absolutely beyond my knowledge/skills. I will work around by simple setting those features with:

```
SET(CMAKE_CXX_COMPILE_FEATURES "cxx_nullptr;cxx_override")
```

Hopefully it's possible to add proper compiler-feature detection with cmake in the future. Also I have no idea where to report if conda-build is the wrong place.

**Comment by looooo** [comment]
next problem: zlib in netgen-feedstock is used from the build-environment 👎

**Comment by mingwandroid** [comment]
> Hopefully it's possible to add proper compiler-feature detection with cmake in the future. Also I have no idea where to report if conda-build is the wrong place

Since you are running CMake here, Occams Razor would suggest this is an issue with CMake. But for you to prove that you need to demonstrate that the problem happens outside of the Anaconda Distribution (so official CMake, official Apple compilers). If it doesn't reproduce there then it's either a problem with our CMake packages, our compiler packages or an interaction between the two. Either way, it's got nothing to do with conda-build!

**Comment by looooo** [comment]
>  But for you to prove that you need to demonstrate that the problem happens outside of the Anaconda Distribution (so official CMake, official Apple compilers).

I guess it will work with official compilers. Searching for this error gave some reports of problems with feature-detection with cmake-minimum-requirement < 3.3. But with setting the cmake-policy `CMP0025` to  NEW, this was resolved. So my guess is that this problem is specific to conda. If it is caused by cmake or the compilers, I don't know.

**Comment by mingwandroid** [comment]
Please stop guessing. Investigate. You will learn that way.

**Comment by looooo** [comment]
> Please stop guessing. Investigate. You will learn that way.

Here is the minimal example. No problems with official compilers. Feel free to try it yourself:
https://github.com/looooo/cmake_tests

```
The C compiler identification is AppleClang 10.0.0.10001145
The CXX compiler identification is AppleClang 10.0.0.10001145
Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Detecting C compile features
Detecting C compile features - done
Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Detecting CXX compile features
Detecting CXX compile features - done
Configuring done
Generating done
```

But problems with conda:
```
-- The C compiler identification is Clang 4.0.1
-- The CXX compiler identification is Clang 4.0.1
-- Check for working C compiler: $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang
-- Check for working C compiler: $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++
-- Check for working CXX compiler: $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
CMake Error in CMakeLists.txt:
  No known features for CXX compiler

  "Clang"

  version 4.0.1.


```

**Comment by mingwandroid** [comment]
Thanks for doing that. I'm in the process of making Cmake work better within the ecosystem BTW.

**Comment by isuruf** [comment]
@looooo, do you have the CMakeError.log and CMakeOutput.log for the small project?

**Comment by isuruf** [comment]
Related to https://gitlab.kitware.com/cmake/cmake/issues/17712

Need the `cmake_minimum_required` before the `project` command. Try swapping the first 2 lines of your example project.

**Comment by isuruf** [comment]
From, https://cmake.org/cmake/help/v3.10/command/cmake_minimum_required.html

> Note

> Call the cmake_minimum_required() command at the beginning of the top-level CMakeLists.txt file even before calling the project() command. It is important to establish version and policy settings before invoking other commands whose behavior they may affect.

**Comment by looooo** [comment]
@isuruf Thanks for the information. Great, this really solves this issue!

**Pull request merged**